crates/api/src/site/registration_applications/approve.rs

```diff
@@ -25,41 +29,66 @@ pub async fn approve_registration_application(
-  };
+  let pool_context = context.clone();
+  let pool = &mut pool_context.pool();
+  let conn = &mut get_conn(pool).await?;
+  let (approval_email_user, result) = conn
+    .build_transaction()
+    .run(|conn| {
+      Box::pin(async move {
+        // Update the registration with reason, admin_id
+        let deny_reason = diesel_string_update(data.deny_reason.as_deref());
+        let app_form = RegistrationApplicationUpdateForm {
+          admin_id: Some(Some(local_user_view.person.id)),
+          deny_reason,
+        };
 
-  let registration_application =
-    RegistrationApplication::update(&mut context.pool(), app_id, &app_form).await?;
+        let registration_application =
+          RegistrationApplication::update(&mut conn.into(), app_id, &app_form).await?;
 
-  // Update the local_user row
-  let local_user_form = LocalUserUpdateForm {
-    accepted_application: Some(data.approve),
-    ..Default::default()
-  };
+        // Update the local_user row
+        let local_user_form = LocalUserUpdateForm {
+          accepted_application: Some(data.approve),
+          ..Default::default()
+        };
 
-  let approved_user_id = registration_application.local_user_id;
-  LocalUser::update(&mut context.pool(), approved_user_id, &local_user_form).await?;
+        let approved_user_id = registration_application.local_user_id;
+        LocalUser::update(&mut conn.into(), approved_user_id, &local_user_form).await?;
 
-  if data.approve {
-    let approved_local_user_view = LocalUserView::read(&mut context.pool(), approved_user_id)
-      .await?
-      .ok_or(LemmyErrorType::CouldntFindLocalUser)?;
+        // This view is only Some if an email should be sent
+        let approval_email_user_view = match data.approve {
+          true => {
+            let approved_local_user_view = LocalUserView::read(&mut conn.into(), approved_user_id)
+              .await?
+              .ok_or(LemmyErrorType::CouldntFindLocalUser)?;
+            match approved_local_user_view.local_user.email.is_some() {
+              true => Some(approved_local_user_view),
+              false => None,
+            }
+          }
+          false => None,
+        };
 
-    if approved_local_user_view.local_user.email.is_some() {
-      send_application_approved_email(&approved_local_user_view, context.settings()).await?;
-    }
-  }
+        // Read the view
+        let registration_application = RegistrationApplicationView::read(&mut conn.into(), app_id)
+          .await?
+          .ok_or(LemmyErrorType::CouldntFindRegistrationApplication)?;
 
-  // Read the view
-  let registration_application = RegistrationApplicationView::read(&mut context.pool(), app_id)
-    .await?
-    .ok_or(LemmyErrorType::CouldntFindRegistrationApplication)?;
+        Ok::<_, LemmyError>((
+          approval_email_user_view,
+          Json(RegistrationApplicationResponse {
+            registration_application,
+          }),
+        ))
+      }) as _
+    })
+    .await?;
```

Comment:
Moving the things other than updates into the transaction doesn't remove the possibility of inconsistency, and I don't know if that's necessary.

Reply:
What do you mean?

The transaction spans from the first SQL query to the last one, so if anything in the middle fails the transaction gets rolled back.

Reply:
got it

Reply:
should be good now